### PR TITLE
Make it unsafe to define NMI handlers

### PR DESCRIPTION
**Pull request by jonas-schievink**
Reverts https://github.com/rust-embedded/cortex-m-rt/pull/257

Fixes https://github.com/rust-embedded/cortex-m-rt/issues/269
Fixes https://github.com/rust-embedded/cortex-m/issues/196 (see that thread for details)

### CODE REVIEW

File: macros/src/lib.rs [hunk 1]
```diff
@@ -130,28 +138,38 @@ pub fn exception(args: TokenStream, input: TokenStream) -> TokenStream {
...
+    if f.sig.unsafety.is_none() {
+        match exn {
+            Exception::DefaultHandler | Exception::HardFault | Exception::NonMaskableInt => {
+                // These are unsafe to define.
+                let name = if exn == Exception::DefaultHandler {
+                    format!("`DefaultHandler`")
+                } else {
+                    format!("`{:?}` handler", exn)
+                };
+                return parse::Error::new(ident.span(), format_args!("defining a {} is unsafe and requires an `unsafe fn` (see the cortex-m-rt docs)", name))
```
Comment (jonas-schievink) [inline macros/src/lib.rs:166]: Oh the docs are actually still missing, hang on

**Review by therealprof — APPROVED** [review]
LGTM

**Review by therealprof — CHANGES_REQUESTED** [review]
Uhm, just noticed that this is missing a CHANGELOG entry.

**Comment by jonas-schievink** [comment]
We're missing *all the* changelog entries. I'll add one to https://github.com/rust-embedded/cortex-m-rt/pull/288 to avoid merge conflicts.

**Review by therealprof — APPROVED** [review]
Right, but the plan was to keep the CHANGELOG entries together with the changes in the future.

bors+

**Comment by thalesfragoso** [comment]
bors r=therealprof

**Pull request merged**